### ISSUE TITLE
Tests on `char` broken on riscv64 and sparc64

### ISSUE THREAD

**Issue by hack3ric** [rkyv/rend#8]
Replace `mem::transmute::<_, [u8; 4]>(char_{be,le}::new('🎉'))` with `u32::from('🎉').to_{be,le}_bytes()` and the assert passes on riscv64. I don't have a sparc64 environment to test on.

From Debian buildd log: [riscv64](https://buildd.debian.org/status/fetch.php?pkg=rust-rend&arch=riscv64&ver=0.3.6-1&stamp=1657488217&raw=0) [sparc64](https://buildd.debian.org/status/fetch.php?pkg=rust-rend&arch=sparc64&ver=0.3.6-1&stamp=1660277878&raw=0)

riscv64:
```
running 1 test
test tests::endian_representation ... FAILED

failures:

---- tests::endian_representation stdout ----
thread 'tests::endian_representation' panicked at 'assertion failed: `(left == right)`
  left: `[0, 1, 243, 137]`,
 right: `[0, 1, 243, 0]`', src/lib.rs:466:13
stack backtrace:
   0: rust_begin_unwind
             at /usr/src/rustc-1.59.0/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /usr/src/rustc-1.59.0/library/core/src/panicking.rs:116:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /usr/src/rustc-1.59.0/library/core/src/panicking.rs:154:5
   4: rend::tests::endian_representation
             at /usr/share/cargo/registry/rend-0.3.6/src/lib.rs:466:13
   5: rend::tests::endian_representation::{{closure}}
             at /usr/share/cargo/registry/rend-0.3.6/src/lib.rs:351:5
   6: core::ops::function::FnOnce::call_once
             at /usr/src/rustc-1.59.0/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /usr/src/rustc-1.59.0/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    tests::endian_representation

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.08s

error: test failed, to rerun pass '--lib'
```

sparc64:
```
running 1 test
test tests::endian_representation ... FAILED

failures:

---- tests::endian_representation stdout ----
thread 'tests::endian_representation' panicked at 'assertion failed: `(left == right)`
  left: `[137, 243, 1, 0]`,
 right: `[0, 243, 1, 0]`', src/lib.rs:470:13
stack backtrace:
   0: rust_begin_unwind
             at /usr/src/rustc-1.59.0/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /usr/src/rustc-1.59.0/library/core/src/panicking.rs:116:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /usr/src/rustc-1.59.0/library/core/src/panicking.rs:154:5
   4: rend::tests::endian_representation
             at /usr/share/cargo/registry/rend-0.3.6/src/lib.rs:470:13
   5: rend::tests::endian_representation::{{closure}}
             at /usr/share/cargo/registry/rend-0.3.6/src/lib.rs:351:5
   6: core::ops::function::FnOnce::call_once
             at /usr/src/rustc-1.59.0/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /usr/src/rustc-1.59.0/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    tests::endian_representation

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s

error: test failed, to rerun pass '--lib'
```

**Comment by djkoloski** [comment]
This looks like an issue with the way that `char`s of non-native endianness are stored. During the test, an invalid char is created between construction and evaluation which is likely the source of this issue. To fix this, we'll have to change `LittleEndian<char>` and `BigEndian<char>` to store the underlying value as a `u32`. Unfortunately, this will also inhibit any niching optimizations.

**Comment by nc7s** [comment]
Ran some tests, see below:

```rust
fn main() {
	let c = '🎉';
	dbg!(
	unsafe { (
		std::mem::transmute::<_, [u8; 4]>(
				c as u32
		),
		std::mem::transmute::<_, [u8; 4]>(
			core::char::from_u32_unchecked(
				c as u32
			)
		),
		std::mem::transmute::<_, [u8; 4]>(
				(c as u32).swap_bytes()
		),
		std::mem::transmute::<_, [u8; 4]>(
			core::char::from_u32_unchecked(
				(c as u32).swap_bytes()
			)
		),
		std::mem::transmute::<_, [u8; 4]>(
				(c as u32).swap_bytes().swap_bytes()
		),
		std::mem::transmute::<_, [u8; 4]>(
			core::char::from_u32_unchecked(
				(c as u32).swap_bytes().swap_bytes()
			)
		),
	) }
	);
}

```

The output on arm64 is:

```
[test.rs:3] unsafe {
    (std::mem::transmute::<_, [u8; 4]>(c as u32),
        std::mem::transmute::<_,
                [u8; 4]>(core::char::from_u32_unchecked(c as u32)),
        std::mem::transmute::<_, [u8; 4]>((c as u32).swap_bytes()),
        std::mem::transmute::<_,
                [u8; 4]>(core::char::from_u32_unchecked((c as
                            u32).swap_bytes())),
        std::mem::transmute::<_,
                [u8; 4]>((c as u32).swap_bytes().swap_bytes()),
        std::mem::transmute::<_,
                [u8; 4]>(core::char::from_u32_unchecked((c as
                                u32).swap_bytes().swap_bytes())))
} = (
    [ 137, 243, 1, 0, ],
    [ 137, 243, 1, 0, ],
    [ 0, 1, 243, 137, ],
    [ 0, 1, 243, 137, ],
    [ 137, 243, 1, 0, ],
    [ 137, 243, 1, 0, ],
)
```

But on riscv64 it's:

```
[test.rs:3] unsafe {
    (std::mem::transmute::<_, [u8; 4]>(c as u32),
     std::mem::transmute::<_,
                           [u8; 4]>(core::char::from_u32_unchecked(c as u32)),
     std::mem::transmute::<_, [u8; 4]>((c as u32).swap_bytes()),
     std::mem::transmute::<_,
                           [u8; 4]>(core::char::from_u32_unchecked((c as
                                                                        u32).swap_bytes())),
     std::mem::transmute::<_, [u8; 4]>((c as u32).swap_bytes().swap_bytes()),
     std::mem::transmute::<_,
                           [u8; 4]>(core::char::from_u32_unchecked((c as
                                                                        u32).swap_bytes().swap_bytes())))
} = (
    [ 137, 243, 1, 0, ],
    [ 137, 243, 1, 0, ],
    [ 0, 1, 243, 137, ],
    [ 0, 1, 243, 0, ],
    [ 137, 243, 1, 0, ],
    [ 137, 243, 1, 0, ],
)
```

The 4th case, `u32::swap_bytes()` then `char::from_u32_unchecked()`, missed the 4th byte, but the double swapped case didn't.

Environments:

1. rustc installed through rustup on macOS arm64
  rustc 1.63.0 (4b91a6ea7 2022-08-08)
  binary: rustc
  commit-hash: 4b91a6ea7258a947e59c6522cd5898e7c0a6a88f
  commit-date: 2022-08-08
  host: aarch64-apple-darwin
  release: 1.63.0
  LLVM version: 14.0.5

2. rustc installed through unstable on Debian, chrooted qemu-user-riscv64 on amd64
rustc 1.59.0 # 1.59.0+dfsg1-2
binary: rustc
commit-hash: unknown
commit-date: unknown
host: riscv64gc-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.1

**Comment by djkoloski** [comment]
Thanks for running these tests, they confirm my suspicions. I think the best approach would be to introduce a new trait with an associated type that defines the storage type for `LittleEndian` and `BigEndian` on non-native platforms. For `char`, this would be `u32` to avoid passing through a `char` and causing UB. This will involve new bounds on the type parameter and so will require a version bump.

**Comment by djkoloski** [comment]
Fixed by 6f75ee60b1d10914bf045cf04c72749552d05800, tested on `riscv64gc-unknown-linux-gnu` through `cargo cross`.

**Comment by djkoloski** [comment]
This is now released in version 0.4.0.
